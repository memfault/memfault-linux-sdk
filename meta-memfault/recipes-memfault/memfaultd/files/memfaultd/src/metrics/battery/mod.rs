//
// Copyright (c) Memfault, Inc.
// See License.txt for details
mod battery_reading_handler;
pub use battery_reading_handler::BatteryReadingHandler;

mod battery_monitor;
pub use battery_monitor::BatteryMonitor;
pub use battery_monitor::BatteryMonitorReading;
