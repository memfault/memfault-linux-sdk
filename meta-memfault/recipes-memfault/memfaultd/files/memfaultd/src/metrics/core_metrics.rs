//
// Copyright (c) Memfault, Inc.
// See License.txt for details
// Connectivity metrics
pub const METRIC_MF_SYNC_SUCCESS: &str = "sync_memfault_successful";
pub const METRIC_MF_SYNC_FAILURE: &str = "sync_memfault_failure";
pub const METRIC_CONNECTED_TIME: &str = "connectivity_connected_time_ms";
pub const METRIC_EXPECTED_CONNECTED_TIME: &str = "connectivity_expected_time_ms";
pub const METRIC_SYNC_SUCCESS: &str = "sync_successful";
pub const METRIC_SYNC_FAILURE: &str = "sync_failure";

// Stability metrics
pub const METRIC_OPERATIONAL_HOURS: &str = "operational_hours";
pub const METRIC_OPERATIONAL_CRASHFREE_HOURS: &str = "operational_crashfree_hours";
pub const METRIC_OPERATIONAL_CRASHES: &str = "operational_crashes";

// Battery metrics
pub const METRIC_BATTERY_DISCHARGE_DURATION_MS: &str = "battery_discharge_duration_ms";
pub const METRIC_BATTERY_SOC_PCT_DROP: &str = "battery_soc_pct_drop";
pub const METRIC_BATTERY_SOC_PCT: &str = "battery_soc_pct";
