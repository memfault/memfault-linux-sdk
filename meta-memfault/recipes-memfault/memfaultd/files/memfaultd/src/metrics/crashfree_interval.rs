//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{
    cmp::max,
    iter::once,
    sync::mpsc::{channel, Receiver, Sender},
    time::Duration,
};

use chrono::Utc;
use tiny_http::{Method, Request, Response};

use crate::{
    http_server::{HttpHandler, HttpHandlerResult},
    metrics::{
        core_metrics::{
            METRIC_OPERATIONAL_CRASHES, METRIC_OPERATIONAL_CRASHFREE_HOURS,
            METRIC_OPERATIONAL_HOURS,
        },
        MetricReading,
    },
    util::time_measure::TimeMeasure,
};

use super::{KeyedMetricReading, MetricStringKey};

pub struct CrashFreeIntervalTracker<T: TimeMeasure> {
    last_interval_mark: T,
    last_crashfree_interval_mark: T,
    crash_count: u32,
    sender: Sender<T>,
    receiver: Receiver<T>,
    interval: Duration,
    elapsed_intervals_key: MetricStringKey,
    crashfree_intervals_key: MetricStringKey,
    crash_count_key: MetricStringKey,
}

#[derive(Debug, PartialEq, Eq)]
struct TimeMod<T: TimeMeasure> {
    count: u32,
    mark: T,
}

impl<T> CrashFreeIntervalTracker<T>
where
    T: TimeMeasure + Copy + Ord + std::ops::Add<Duration, Output = T> + Send + Sync + 'static,
{
    pub fn new(
        interval: Duration,
        elapsed_intervals_key: MetricStringKey,
        crashfree_intervals_key: MetricStringKey,
        crash_count_key: MetricStringKey,
    ) -> Self {
        let (sender, receiver) = channel();
        Self {
            last_crashfree_interval_mark: T::now(),
            last_interval_mark: T::now(),
            sender,
            receiver,
            crash_count: 0,
            interval,
            elapsed_intervals_key,
            crashfree_intervals_key,
            crash_count_key,
        }
    }

    /// Returns a tracker with an hourly interval
    pub fn new_hourly() -> Self {
        Self::new(
            Duration::from_secs(3600),
            METRIC_OPERATIONAL_HOURS.parse().unwrap(),
            METRIC_OPERATIONAL_CRASHFREE_HOURS.parse().unwrap(),
            METRIC_OPERATIONAL_CRASHES.parse().unwrap(),
        )
    }

    /// Wait for the next crash or update the metrics if the wait duration has passed.
    ///
    /// This allows us to have instant updates on crashes and hourly updates on the metrics, but
    /// also allows us to periodically update the metrics so that we don't have to wait for a crash.
    pub fn wait_and_update(&mut self, wait_duration: Duration) -> Vec<KeyedMetricReading> {
        if let Ok(crash_ts) = self.receiver.recv_timeout(wait_duration) {
            // Drain the receiver to get all crashes that happened since the last update
            self.receiver
                .try_iter()
                .chain(once(crash_ts))
                .for_each(|ts| {
                    self.crash_count += 1;
                    self.last_crashfree_interval_mark = max(self.last_crashfree_interval_mark, ts);
                });
        }

        // Since timing out just means no crashes occurred in the `wait_duration`,
        // update even when the receiver times out.
        self.update()
    }

    fn update(&mut self) -> Vec<KeyedMetricReading> {
        let TimeMod {
            count: count_op_interval,
            mark: last_counted_op_interval,
        } = Self::full_interval_elapsed_since(self.interval, &self.last_interval_mark);
        let TimeMod {
            count: count_crashfree_interval,
            mark: last_counted_crashfree_interval,
        } = Self::full_interval_elapsed_since(self.interval, &self.last_crashfree_interval_mark);

        self.last_interval_mark = last_counted_op_interval;
        self.last_crashfree_interval_mark = last_counted_crashfree_interval;

        let crashes = self.crash_count;
        self.crash_count = 0;

        let metrics_ts = Utc::now();
        vec![
            KeyedMetricReading::new(
                self.elapsed_intervals_key.clone(),
                MetricReading::Counter {
                    value: count_op_interval as f64,
                    timestamp: metrics_ts,
                },
            ),
            KeyedMetricReading::new(
                self.crashfree_intervals_key.clone(),
                MetricReading::Counter {
                    value: count_crashfree_interval as f64,
                    timestamp: metrics_ts,
                },
            ),
            KeyedMetricReading::new(
                self.crash_count_key.clone(),
                MetricReading::Counter {
                    value: crashes as f64,
                    timestamp: metrics_ts,
                },
            ),
        ]
    }

    pub fn http_handler(&mut self) -> Box<dyn HttpHandler> {
        Box::new(CrashFreeIntervalHttpHandler {
            channel: self.sender.clone(),
        })
    }

    pub fn capture_crash(&self) {
        self.sender
            .send(T::now())
            .expect("Failed to send crash timestamp");
    }

    /// Count how many `interval` have elapsed since `since`.
    ///
    /// This returns the number of intervals that have elapsed since `since`, and the timestamp of the end of the last interval
    /// that was counted. This is the value you should pass as `since` next time you call this function.
    ///
    /// See unit test for examples.
    fn full_interval_elapsed_since(interval: Duration, since: &T) -> TimeMod<T> {
        let now = T::now();
        if *since > now {
            return TimeMod {
                count: 0,
                mark: T::now(),
            };
        }

        let duration = now.since(since);
        let count_interval_elapsed = (duration.as_nanos() / interval.as_nanos()) as u32;
        TimeMod {
            count: count_interval_elapsed,
            mark: since.add(interval * count_interval_elapsed),
        }
    }
}

struct CrashFreeIntervalHttpHandler<T> {
    channel: Sender<T>,
}

impl<T> HttpHandler for CrashFreeIntervalHttpHandler<T>
where
    T: TimeMeasure + Copy + Ord + std::ops::Add<Duration, Output = T> + Send + Sync,
{
    fn handle_request(&self, request: &mut Request) -> HttpHandlerResult {
        if request.url() == "/v1/crash/report" && request.method() == &Method::Post {
            self.channel.send(T::now()).unwrap();
            HttpHandlerResult::Response(Response::from_string("OK").boxed())
        } else {
            HttpHandlerResult::NotHandled
        }
    }
}

#[cfg(test)]
mod tests {
    use std::time::Duration;

    use rstest::rstest;

    use crate::{
        metrics::{
            crashfree_interval::{
                METRIC_OPERATIONAL_CRASHES, METRIC_OPERATIONAL_CRASHFREE_HOURS,
                METRIC_OPERATIONAL_HOURS,
            },
            KeyedMetricReading, MetricReading,
        },
        test_utils::TestInstant,
    };

    use super::CrashFreeIntervalTracker;
    use super::TimeMod;

    #[rstest]
    fn test_counting_intervals() {
        use std::time::Duration;

        // move the clock forward so we can go backwards below
        TestInstant::sleep(Duration::from_secs(3600));
        let now = TestInstant::now();

        let d10 = Duration::from_secs(10);
        assert_eq!(
            CrashFreeIntervalTracker::full_interval_elapsed_since(d10, &now),
            TimeMod {
                count: 0,
                mark: now
            }
        );
        assert_eq!(
            CrashFreeIntervalTracker::full_interval_elapsed_since(
                d10,
                &(now - Duration::from_secs(10))
            ),
            TimeMod {
                count: 1,
                mark: now
            }
        );
        assert_eq!(
            CrashFreeIntervalTracker::full_interval_elapsed_since(
                d10,
                &(now - Duration::from_secs(25))
            ),
            TimeMod {
                count: 2,
                mark: now - Duration::from_secs(5)
            }
        );
    }

    #[rstest]
    fn test_counting_hours() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new_hourly();

        TestInstant::sleep(Duration::from_secs(7200));

        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            2,
            2,
        );
    }

    #[rstest]
    fn test_counting_minutes() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new(
            Duration::from_secs(60),
            METRIC_OPERATIONAL_HOURS.parse().unwrap(),
            METRIC_OPERATIONAL_CRASHFREE_HOURS.parse().unwrap(),
            METRIC_OPERATIONAL_CRASHES.parse().unwrap(),
        );

        TestInstant::sleep(Duration::from_secs(3600));
        crashfree_tracker.capture_crash();
        TestInstant::sleep(Duration::from_secs(3600));

        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            120,
            60,
        );
    }

    #[rstest]
    fn test_30min_heartbeat() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new_hourly();

        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            0,
            0,
        );

        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            1,
            1,
        );
    }

    #[rstest]
    fn test_30min_heartbeat_with_crash() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new_hourly();

        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            0,
            0,
        );

        // Crash at t0 + 30min
        crashfree_tracker.capture_crash();

        // After 30' we should be ready to mark an operational hour
        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            1,
            0,
        );

        // After another 30' we should be ready to mark another crashfree hour
        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            0,
            1,
        );

        // After another 30' we should be ready to mark another operational hour
        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            1,
            0,
        );
    }

    #[rstest]
    fn test_180min_heartbeat_with_one_crash() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new_hourly();

        // Basic test
        TestInstant::sleep(Duration::from_secs(3600 * 3));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            3,
            3,
        );

        // Crash at interval + 170'
        TestInstant::sleep(Duration::from_secs(170 * 60));
        crashfree_tracker.capture_crash();

        // Another 10' to the heartbeat mark
        // We will count 0 operational hour here. That is a consequence of the heartbeat being larger than the hour
        // To avoid this bug, we need to make sure we call the `update` at least once per hour!
        TestInstant::sleep(Duration::from_secs(10 * 60));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            3,
            0,
        );

        // However, doing the crash at interval +10' then waiting for 170' will record 2 crashfree hours
        TestInstant::sleep(Duration::from_secs(10 * 60));
        crashfree_tracker.capture_crash();
        TestInstant::sleep(Duration::from_secs(170 * 60));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            3,
            2,
        );
    }

    fn assert_operational_metrics(
        metrics: Vec<KeyedMetricReading>,
        expected_op_hours: u32,
        expected_crashfree_hours: u32,
    ) {
        assert_eq!(metrics.len(), 3);
        let op_hours = metrics
            .iter()
            .find(|m| m.name.as_str() == METRIC_OPERATIONAL_HOURS)
            .unwrap();
        let crash_free_hours = metrics
            .iter()
            .find(|m| m.name.as_str() == METRIC_OPERATIONAL_CRASHFREE_HOURS)
            .unwrap();

        let op_hours_value = match op_hours.value {
            MetricReading::Counter { value, .. } => value,
            _ => panic!("Unexpected metric type"),
        };

        let crashfree_hours_value = match crash_free_hours.value {
            MetricReading::Counter { value, .. } => value,
            _ => panic!("Unexpected metric type"),
        };

        assert_eq!(
            (op_hours_value as u32, crashfree_hours_value as u32),
            (expected_op_hours, expected_crashfree_hours)
        );
    }
}
