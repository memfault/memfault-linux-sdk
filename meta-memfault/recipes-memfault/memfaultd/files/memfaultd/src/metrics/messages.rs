//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::path::PathBuf;

use eyre::Result;
use ssf::{Message, MsgMailbox};

use crate::network::NetworkConfig;

use super::{KeyedMetricReading, SessionName};

/// Allows KeyedMetricReading to be sent as a message. The `ssf` framework will
/// automatically support sending `Vec<KeyedMetricReading>` as well.
impl Message for KeyedMetricReading {
    type Reply = Result<()>;
}

/// Syntactic Sugar because this is the most-used type of mailbox in memfaultd.
pub type MetricsMBox = MsgMailbox<Vec<KeyedMetricReading>>;

pub enum SessionEventMessage {
    StartSession {
        name: SessionName,
        readings: Vec<KeyedMetricReading>,
    },
    StopSession {
        name: SessionName,
        readings: Vec<KeyedMetricReading>,
        mar_staging_area: PathBuf,
        network_config: NetworkConfig,
    },
}

impl Message for SessionEventMessage {
    type Reply = Result<()>;
}
