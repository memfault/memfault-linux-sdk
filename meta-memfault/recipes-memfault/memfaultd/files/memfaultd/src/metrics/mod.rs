//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use chrono::{DateTime, Utc};

mod battery;
pub use battery::BatteryMonitor;
pub use battery::BatteryMonitorReading;
pub use battery::BatteryReadingHandler;

mod connectivity;
pub use connectivity::ConnectivityMonitor;
pub use connectivity::ReportSyncEventHandler;

mod metric_string_key;
pub use metric_string_key::MetricStringKey;

mod metric_report;
pub use metric_report::MetricReport;
pub use metric_report::MetricReportType;

mod metric_report_manager;
pub use metric_report_manager::MetricReportManager;
pub use metric_report_manager::TakeMetrics;

mod messages;
pub use messages::*;

mod metric_reading;
pub use metric_reading::KeyedMetricReading;
pub use metric_reading::MetricReading;

mod timeseries;

mod metric_value;
pub use metric_value::MetricValue;

pub type MetricTimestamp = DateTime<Utc>;

mod periodic_metric_report;
pub use periodic_metric_report::PeriodicMetricReportDumper;

mod crashfree_interval;
pub use crashfree_interval::CrashFreeIntervalTracker;

mod session_name;
pub use session_name::SessionName;

mod session_event_handler;
pub use session_event_handler::SessionEventHandler;

pub mod core_metrics;

pub mod statsd_server;
pub use statsd_server::StatsDServer;

mod system_metrics;
pub use system_metrics::SystemMetricsCollector;
pub use system_metrics::BUILTIN_SYSTEM_METRIC_NAMESPACES;
