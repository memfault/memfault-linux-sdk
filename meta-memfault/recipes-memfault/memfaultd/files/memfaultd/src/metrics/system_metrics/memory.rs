//
// Copyright (c) Memfault, Inc.
// See License.txt for details
//! Collect memory metric readings from /proc/meminfo
//!
//! This module parses memory statistics from /proc/meminfo and
//! constructs KeyedMetricReadings based on those statistics.
//!
//! Example /proc/meminfo contents:
//! MemTotal:         365916 kB
//! MemFree:          242276 kB
//! MemAvailable:     292088 kB
//! Buffers:            4544 kB
//! Cached:            52128 kB
//! SwapCached:            0 kB
//! Active:            21668 kB
//! Inactive:          51404 kB
//! Active(anon):       2312 kB
//! Inactive(anon):    25364 kB
//! Active(file):      19356 kB
//! Inactive(file):    26040 kB
//! Unevictable:        3072 kB
//! Mlocked:               0 kB
//! SwapTotal:             0 kB
//! SwapFree:              0 kB
//! Dirty:                 0 kB
//! Writeback:             0 kB
//! AnonPages:         19488 kB
//! Mapped:            29668 kB
//! Shmem:             11264 kB
//! KReclaimable:      14028 kB
//! Slab:              32636 kB
//! SReclaimable:      14028 kB
//!
//! Only the following lines are currently processed:
//! MemTotal, MemFree, and optionally MemAvailable
//!
//! These lines are used by this module to calculate
//! free and used memory. MemFree is used in place of
//! MemAvailable if the latter is not present.
//!
//!
//! See additional Linux kernel documentation on /proc/meminfo here:
//! https://www.kernel.org/doc/html/latest/filesystems/proc.html#meminfo
use std::fs::read_to_string;
use std::path::Path;
use std::{collections::HashMap, str::FromStr};

use eyre::{eyre, Result};
use nom::{
    bytes::complete::tag,
    character::complete::{alpha1, multispace1},
    number::complete::double,
    sequence::{delimited, terminated},
    IResult,
};

use crate::metrics::{
    core_metrics::METRIC_MEMORY_PCT, system_metrics::SystemMetricFamilyCollector,
    KeyedMetricReading, MetricStringKey,
};

const PROC_MEMINFO_PATH: &str = "/proc/meminfo";
pub const MEMORY_METRIC_NAMESPACE: &str = "memory";

pub struct MemoryMetricsCollector;

impl MemoryMetricsCollector {
    pub fn new() -> Self {
        MemoryMetricsCollector {}
    }

    pub fn get_memory_metrics(&self) -> Result<Vec<KeyedMetricReading>> {
        let path = Path::new(PROC_MEMINFO_PATH);
        // Need to read all of /proc/meminfo at once
        // as we derive used memory based on a calculation
        // using multiple lines
        let contents = read_to_string(path)?;
        Self::parse_meminfo_memory_stats(&contents)
    }

    /// Parses the key in a /proc/meminfo line
    ///
    /// A key is the string terminated by the `:` character
    /// In the following line, "MemTotal" would be parsed as the key
    /// MemTotal:         365916 kB
    fn parse_meminfo_key(meminfo_line: &str) -> IResult<&str, &str> {
        terminated(alpha1, tag(":"))(meminfo_line)
    }

    /// Parses the kilobyte value in a /proc/meminfo line
    ///
    /// This value is the kilobytes used by the corresponding
    /// meminfo key. The value is a number terminated by " kB".
    /// In the following line, 365916.0 would be parsed by
    /// this function as the kB used by "MemTotal"
    /// MemTotal:         365916 kB
    fn parse_meminfo_kb(meminfo_line_suffix: &str) -> IResult<&str, f64> {
        delimited(multispace1, double, tag(" kB"))(meminfo_line_suffix)
    }

    /// Parses a full /proc/meminfo contents and returns
    /// a vector of KeyedMetricReadings
    fn parse_meminfo_memory_stats(meminfo: &str) -> Result<Vec<KeyedMetricReading>> {
        let mut stats = meminfo
            .trim()
            .lines()
            .map(|line| -> Result<(&str, f64), String> {
                let (suffix, key) = Self::parse_meminfo_key(line).map_err(|e| e.to_string())?;
                let (_, kb) = Self::parse_meminfo_kb(suffix).map_err(|e| e.to_string())?;
                // Use bytes as unit instead of KB
                Ok((key, kb * 1024.0))
            })
            .filter_map(|result| result.ok())
            .collect::<HashMap<&str, f64>>();

        // Use the same methodology as `free` to calculate used memory.
        //
        // For kernels 3.14 and greater:
        // MemUsed = MemTotal - MemAvailable
        //
        // For kernels less than 3.14 (no MemAvailable):
        // MemUsed = MemTotal - MemFree
        //
        // See below man page for more info:
        // https://man7.org/linux/man-pages/man1/free.1.html
        let total = stats
            .remove("MemTotal")
            .ok_or_else(|| eyre!("{} is missing required value MemTotal", PROC_MEMINFO_PATH))?;
        let free = stats
            .remove("MemFree")
            .ok_or_else(|| eyre!("{} is missing required value MemFree", PROC_MEMINFO_PATH))?;

        // Check that MemTotal is nonzero to avoid dividing by 0
        if total != 0.0 {
            let available = stats.remove("MemAvailable").unwrap_or(free);

            let used = total - available;

            let pct_used = (used / total) * 100.0;

            let used_key = MetricStringKey::from_str("memory/memory/used")
                .map_err(|e| eyre!("Failed to construct MetricStringKey for used memory: {}", e))?;
            let free_key = MetricStringKey::from_str("memory/memory/free")
                .map_err(|e| eyre!("Failed to construct MetricStringKey for used memory: {}", e))?;

            let pct_key = MetricStringKey::from_str(METRIC_MEMORY_PCT)
                .map_err(|e| eyre!("Failed to construct MetricStringKey for used memory: {}", e))?;

            Ok(vec![
                KeyedMetricReading::new_histogram(free_key, free),
                KeyedMetricReading::new_histogram(used_key, used),
                KeyedMetricReading::new_histogram(pct_key, pct_used),
            ])
        } else {
            Err(eyre!("MemTotal is 0, can't calculate memory usage metrics"))
        }
    }
}

impl SystemMetricFamilyCollector for MemoryMetricsCollector {
    fn collect_metrics(&mut self) -> Result<Vec<KeyedMetricReading>> {
        self.get_memory_metrics()
    }

    fn family_name(&self) -> &'static str {
        MEMORY_METRIC_NAMESPACE
    }
}

#[cfg(test)]
mod test {

    use insta::{assert_json_snapshot, rounded_redaction, with_settings};
    use rstest::rstest;

    use super::MemoryMetricsCollector;

    #[rstest]
    #[case("MemTotal:         365916 kB", "MemTotal", 365916.0)]
    #[case("MemFree:          242276 kB", "MemFree", 242276.0)]
    #[case("MemAvailable:     292088 kB", "MemAvailable", 292088.0)]
    #[case("Buffers:            4544 kB", "Buffers", 4544.0)]
    #[case("Cached:            52128 kB", "Cached", 52128.0)]
    fn test_parse_meminfo_line(
        #[case] proc_meminfo_line: &str,
        #[case] expected_key: &str,
        #[case] expected_value: f64,
    ) {
        let (suffix, key) = MemoryMetricsCollector::parse_meminfo_key(proc_meminfo_line).unwrap();
        let (_, kb) = MemoryMetricsCollector::parse_meminfo_kb(suffix).unwrap();

        assert_eq!(key, expected_key);
        assert_eq!(kb, expected_value);
    }

    #[rstest]
    fn test_get_memory_metrics() {
        let meminfo = "MemTotal:         365916 kB
MemFree:          242276 kB
MemAvailable:     292088 kB
Buffers:            4544 kB
Cached:            52128 kB
SwapCached:            0 kB
Active:            21668 kB
Inactive:          51404 kB
Active(anon):       2312 kB
Inactive(anon):    25364 kB
Active(file):      19356 kB
Inactive(file):    26040 kB
Unevictable:        3072 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:                 0 kB
Writeback:             0 kB
AnonPages:         19488 kB
Mapped:            29668 kB
Shmem:             11264 kB
KReclaimable:      14028 kB
        ";

        with_settings!({sort_maps => true}, {
        assert_json_snapshot!(
                              MemoryMetricsCollector::parse_meminfo_memory_stats(meminfo).unwrap(),
                              {"[].value.**.timestamp" => "[timestamp]", "[].value.**.value" => rounded_redaction(5)})
        });
    }

    #[rstest]
    fn test_get_memory_metrics_no_memavailable() {
        let meminfo = "MemTotal:         365916 kB
MemFree:          242276 kB
Buffers:            4544 kB
Cached:            52128 kB
SwapCached:            0 kB
Active:            21668 kB
Inactive:          51404 kB
Active(anon):       2312 kB
Inactive(anon):    25364 kB
Active(file):      19356 kB
Inactive(file):    26040 kB
Unevictable:        3072 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:                 0 kB
Writeback:             0 kB
AnonPages:         19488 kB
Mapped:            29668 kB
Shmem:             11264 kB
KReclaimable:      14028 kB
        ";

        with_settings!({sort_maps => true}, {
        assert_json_snapshot!(
                              MemoryMetricsCollector::parse_meminfo_memory_stats(meminfo).unwrap(),
                              {"[].value.**.timestamp" => "[timestamp]", "[].value.**.value" => rounded_redaction(5)})
        });
    }

    #[rstest]
    fn test_fail_to_parse_bad_meminfo_line() {
        assert!(MemoryMetricsCollector::parse_meminfo_key("MemFree=1080kB").is_err());
        assert!(MemoryMetricsCollector::parse_meminfo_kb("1080 mB").is_err());
    }

    #[rstest]
    fn test_fail_get_metrics_with_missing_required_lines() {
        // MemFree is missing
        let meminfo = "MemTotal:         365916 kB
MemAvailable:     292088 kB
Buffers:            4544 kB
Cached:            52128 kB
SwapCached:            0 kB
Active:            21668 kB
Inactive:          51404 kB
Active(anon):       2312 kB
Inactive(anon):    25364 kB
Active(file):      19356 kB
Inactive(file):    26040 kB
Unevictable:        3072 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:                 0 kB
Writeback:             0 kB
AnonPages:         19488 kB
Mapped:            29668 kB
Shmem:             11264 kB
KReclaimable:      14028 kB
        ";
        assert!(MemoryMetricsCollector::parse_meminfo_memory_stats(meminfo).is_err());
    }

    #[rstest]
    fn test_fail_get_metrics_with_bad_fmt() {
        // Not properly formatted with newlines between each key / kB pair
        let meminfo = "MemTotal:         365916 kB MemFree:          242276 kB
Buffers:            4544 kB Cached:            52128 kB Shmem:             11264 kB
        ";
        assert!(MemoryMetricsCollector::parse_meminfo_memory_stats(meminfo).is_err());
    }

    #[rstest]
    fn test_fail_get_metrics_when_mem_total_is_zero() {
        // MemFree is missing
        let meminfo = "MemTotal:         0 kB
MemAvailable:     292088 kB
Buffers:            4544 kB
Cached:            52128 kB
SwapCached:            0 kB
Active:            21668 kB
Inactive:          51404 kB
Active(anon):       2312 kB
Inactive(anon):    25364 kB
Active(file):      19356 kB
Inactive(file):    26040 kB
Unevictable:        3072 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:                 0 kB
Writeback:             0 kB
AnonPages:         19488 kB
Mapped:            29668 kB
Shmem:             11264 kB
KReclaimable:      14028 kB
        ";
        assert!(MemoryMetricsCollector::parse_meminfo_memory_stats(meminfo).is_err());
    }
}
