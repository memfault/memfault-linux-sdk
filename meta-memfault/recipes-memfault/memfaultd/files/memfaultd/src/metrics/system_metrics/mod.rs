//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::sync::{Arc, Mutex};
use std::thread::{self, spawn};
use std::time::Duration;

use eyre::Result;
use log::warn;

use crate::metrics::MetricReportManager;
mod cpu;
use crate::metrics::system_metrics::cpu::{CpuMetricCollector, CPU_METRIC_NAMESPACE};

mod memory;
use crate::metrics::system_metrics::memory::{MemoryMetricsCollector, MEMORY_METRIC_NAMESPACE};

pub const BUILTIN_SYSTEM_METRIC_NAMESPACES: &[&str; 2] =
    &[CPU_METRIC_NAMESPACE, MEMORY_METRIC_NAMESPACE];

pub struct SystemMetricsCollector {}

impl SystemMetricsCollector {
    pub fn new() -> Self {
        Self {}
    }

    pub fn start(
        &self,
        metric_poll_duration: Duration,
        metric_report_manager: Arc<Mutex<MetricReportManager>>,
    ) -> Result<()> {
        let mut cpu_metric_collector = CpuMetricCollector::new();

        spawn(move || loop {
            thread::sleep(metric_poll_duration);

            match cpu_metric_collector.get_cpu_metrics() {
                Ok(cpu_metrics) => {
                    for metric_reading in cpu_metrics {
                        if let Err(e) = metric_report_manager
                            .lock()
                            .expect("Mutex poisoned")
                            .add_metric(metric_reading)
                        {
                            warn!("Couldn't add CPU metric: {}", e)
                        }
                    }
                }
                Err(e) => warn!("CPU metric collection failed: {}", e),
            }

            match MemoryMetricsCollector::get_memory_metrics() {
                Ok(memory_metric_readings) => {
                    for metric_reading in memory_metric_readings {
                        if let Err(e) = metric_report_manager
                            .lock()
                            .expect("Mutex poisoned")
                            .add_metric(metric_reading)
                        {
                            warn!("Couldn't add memory metric reading: {:?}", e)
                        }
                    }
                }
                Err(e) => warn!("Memory metric collection failed: {:?}", e),
            }
        });

        Ok(())
    }
}

impl Default for SystemMetricsCollector {
    fn default() -> Self {
        Self::new()
    }
}
