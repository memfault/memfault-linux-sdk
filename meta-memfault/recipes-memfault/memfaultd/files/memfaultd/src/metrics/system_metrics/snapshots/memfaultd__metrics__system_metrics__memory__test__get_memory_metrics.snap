---
source: memfaultd/src/metrics/system_metrics/memory.rs
expression: "MemoryMetricsCollector::parse_meminfo_memory_stats(meminfo).unwrap()"
---
[
  {
    "name": "memory/memory/free",
    "value": {
      "Histogram": {
        "value": 248090624.0,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "memory/memory/used",
    "value": {
      "Histogram": {
        "value": 75599872.0,
        "timestamp": "[timestamp]"
      }
    }
  }
]
